### PR TITLE
Fix apply_configs decorator causing function signature to be lost

### PR DESCRIPTION
**Pull request by LeonLuttenberger**
### Feature or Bugfix
- Bugfix
- Refactoring

### Detail
The definition for `apply_configs` so far has been `Callable[..., Any] -> Callable[..., Any]`. This means that IntelliSense as well as MyPy were not able to do check the types whenever a function decorated with `apply_configs` was called. Functions such as `wr.s3.to_parquet` were therefore treated as having a signature of `Callable[..., Any] -> Callable[..., Any]`.

By changing the definition of `apply_configs` to make use of generic types, MyPy was able to find errors related to inconsistent type usages (e.g. `Optional[str]` being passed as `str`). As such, I've had to fix all these type errors as part of this PR.

Furthermore, since making this change, the auto-complete in Visual Studio Code is now able to pick up types for functions such as `wr.s3.to_parquet` again.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 22d26ea61cbc3be920a6be11290af018d8392230
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=09d841ac-0a70-48fc-847a-7e0c35cd30a8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 22d26ea61cbc3be920a6be11290af018d8392230
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=23663f82-6882-4f86-bbe9-55bd01dc7d7e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 22d26ea61cbc3be920a6be11290af018d8392230
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2f0c122e-a762-42f2-8f5c-0d29ecc33171%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 22d26ea61cbc3be920a6be11290af018d8392230
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7ab34f51-c204-43f8-94f5-4c7299b8a57c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by malachi-constant — APPROVED** [review]

File: awswrangler/athena/_utils.py [hunk 1]
```diff
@@ -643,7 +649,7 @@ def describe_table(
 @apply_configs
 def create_ctas_table(  # pylint: disable=too-many-locals
     sql: str,
-    database: str,
+    database: Optional[str] = None,
```
Comment (jaidisido) [inline awswrangler/athena/_utils.py:652]: Nitpick - could you fix it in the docs below too?

**Review by jaidisido — COMMENTED** [review]
Looks good - minor nitpick but plese feel free to merge after fixing it

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: f89c6e3db28d2732796f58220a2288ddfb9a90f2
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=60f0555b-bdf1-4eae-b271-7eed12f7aa19%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 63335b5d2facf09e312b21c7632adbd69fedfc77
* Result: FAILED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=13dc22f6-0d7f-45c3-aeca-4ddf098300f0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Pull request merged**